Widgets: fix linting errors, part 2

jeherve opened the pull request:
Fixes #21993

#### Changes proposed in this Pull Request:

* Widgets: fix linting errors in Mailchimp popup widget.
* Widgets: fix linting errors in gallery widget.
* Widgets: fix linting errors in image widget
* Fix linting errors in community widget
* Fix linting errors in RSS widget
* Fix linting errors in Search widget
* Fix linting errors for old Social Icons widget
* Fix linting errors with the Top Posts widget
* Fix linting errors in Twitter Timeline widget
* fix linting errors in Upcoming Events widget

#### Jetpack product discussion

* N/A

#### Does this pull request change what data or activity we track or use?

* No

#### Testing instructions:

* On a new site, go to Jetpack > Settings, search for widgets, and enable the feature.
* On the same screen, enable the Stats feature if it's not already enabled.
* Go to Plugins > Add New, install and activate the Classic Widgets plugin.
* Go to Appearance > Widgets, and try setting up those widgets.
    * You can use this URL to test the Upcoming Events widget: `https://calendar.google.com/calendar/ical/c_c8svl8hfcru362iflfq1ebnq78%40group.calendar.google.com/public/basic.ics`
* Test in Apperance > Customize and on the frontend as well.

jeherve committed:
Add changelog entry

jeherve committed:
Widgets: fix linting errors in Mailchimp popup widget.

jeherve committed:
Widgets: fix linting errors in gallery widget.

jeherve committed:
Widgets: fix linting errors in image widget

jeherve committed:
Fix linting errors in community widget

jeherve committed:
Fix linting errors in RSS widget

jeherve committed:
Fix linting errors in Search widget

jeherve committed:
Merge remote-tracking branch 'origin/master' into update/phpcs-widgets-2

jeherve committed:
Fix linting errors for old Social Icons widget

jeherve committed:
Fix linting errors with the Top Posts widget

jeherve committed:
Fix linting errors in Twitter Timeline widget

jeherve committed:
fix linting errors in Upcoming Events widget

jeherve commented on projects/plugins/jetpack/modules/widgets/rsslinks-widget.php:
```diff
@@ -169,32 +200,35 @@ function form( $instance ) {
+	 * @param string $type Widget type (posts or comments).
+	 * @param array  $args Widget arguments.
+	 */
+	private function rss_link( $type, $args ) {
```
I refactored that method entirely, to make it easier to translate and somehow easier to parse. The current version was very hard to translate as we had translations split in different strings.

jeherve commented on projects/plugins/jetpack/modules/widgets/social-media-icons.php:
```diff
@@ -1,23 +1,24 @@
  * @extends WP_Widget
+ *
+ * phpcs:disable PEAR.NamingConventions.ValidClassName.Invalid
  */
 class WPCOM_social_media_icons_widget extends WP_Widget {
-
+	// phpcs:enable PEAR.NamingConventions.ValidClassName.Invalid
```
This had to be split instead of just a single `// phpcs:ignore` because otherwise it triggers another phpcs linting error.

jeherve commented on projects/plugins/jetpack/modules/widgets/upcoming-events.php:
```diff
@@ -1,7 +1,28 @@
+ * @package automattic/jetpack
+ */
+
+/**
+ * Register the widget.
+ */
+function upcoming_events_register_widgets() {
+	register_widget( 'Jetpack_Upcoming_Events_Widget' );
+}
+add_action( 'widgets_init', 'upcoming_events_register_widgets' );
```
I moved this to the top of the file to be consistent with the other files.

jeherve committed:
Merge remote-tracking branch 'origin/master' into update/phpcs-widgets-2

anomiex commented on projects/plugins/jetpack/modules/widgets/top-posts.php:
```diff
@@ -246,15 +302,20 @@ function widget( $args, $instance ) {
-		$ordering = isset( $instance['ordering'] ) && 'likes' == $instance['ordering'] ? 'likes' : 'views';
-
-		if ( isset( $instance['display'] ) && in_array( $instance['display'], array( 'grid', 'list', 'text' ) ) ) {
+		$ordering = isset( $instance['ordering'] ) && 'likes' === $instance['ordering'] ?
+			'likes'
+			: 'views';
```
I like better how you did this just above.
```suggestion
		$ordering = isset( $instance['ordering'] ) && 'likes' === $instance['ordering']
			? 'likes'
			: 'views';
```

jeherve replied:
Yep, that was an oversight from me. Committing.

anomiex commented on projects/plugins/jetpack/modules/widgets/top-posts.php:
```diff
@@ -295,9 +356,9 @@ function widget( $args, $instance ) {
+		echo $args['before_widget']; // phpcs:ignore WordPress.Security.EscapeOutput.OutputNotEscaped
 		if ( ! empty( $title ) ) {
-			echo $args['before_title'] . $title . $args['after_title'];
+			echo $args['before_title'] . esc_html( $title ) . $args['after_title']; // phpcs:ignore WordPress.Security.EscapeOutput.OutputNotEscaped
```
(same as above re escaping $title)

anomiex commented on projects/plugins/jetpack/modules/widgets/top-posts.php:
```diff
@@ -681,8 +765,8 @@ public function get_posts( $post_ids, $count, $types = array( 'post', 'page' ) )
 
-			if ( $counter == $count ) {
-				break; // only need to load and show x number of likes
+			if ( $counter == $count ) { // phpcs:ignore WordPress.PHP.StrictComparisons.LooseComparison
```
I wonder if we should just cast `$count` to an integer at the top of the function and make this strict.

Although I see the filter just below for some reason claims `$count` will be a string. Weird.

jeherve replied:
This was one of those cases where I prefer to leave it as it is to avoid breaking something that doesn't immediately seem obvious to me. I think I'll leave it like that for now. :)

anomiex commented on projects/plugins/jetpack/modules/widgets/upcoming-events.php:
```diff
@@ -1,7 +1,28 @@
+ * Register the widget.
+ */
+function upcoming_events_register_widgets() {
+	register_widget( 'Jetpack_Upcoming_Events_Widget' );
```
This nitpick again.
```suggestion
	register_widget( Jetpack_Upcoming_Events_Widget::class );
```

anomiex commented on projects/plugins/jetpack/modules/widgets/upcoming-events.php:
```diff
@@ -26,106 +50,140 @@ function css() {
+		<select id="<?php echo esc_attr( $this->get_field_id( 'count' ) ); ?>" name="<?php echo esc_attr( $this->get_field_name( 'count' ) ); ?>">
 			<?php
 			$i = 1;
 			while ( $i <= 10 ) {
-			?>
-				<option <?php selected( $instance['count'], $i ); ?>><?php echo $i; ?></option>
+				?>
+				<option <?php selected( $instance['count'], $i ); ?>><?php echo (int) $i; ?></option>
 			<?php $i++; } ?>
```
Seems weird that this isn't a for loop.
```suggestion
			<?php for ( $i = 1; $i <= 10; $i++ ) { ?>
				<option <?php selected( $instance['count'], $i ); ?>><?php echo (int) $i; ?></option>
			<?php } ?>
```

jeherve replied:
👍 75bdcd8c991a1da7e475fe980f62c24cdcae3652

anomiex reviewed (commented):
Did some static analysis, but didn't test it myself. Feel free to answer "no" to any of the questions, or otherwise disagree. The only thing that seems like an actual error is the lost handling of `$widgets` being a non-array in two places.

jeherve commented on projects/plugins/jetpack/modules/widgets/migrate-to-core/gallery-widget.php:
```diff
@@ -30,10 +31,13 @@ function jetpack_migrate_gallery_widget() {
+		 * Try to get an unique id for the new type of widget.
+		 * It may be the case that the user has already created a core Gallery Widget
+		 * before the migration begins. (Maybe Jetpack was deactivated during core's upgrade).
+		 */
```
No, that was me, I personally find this more readable that way, but I realize that's very much a personal thing. :)

anomiex committed:
avoid warnings when widgets is not an array

See https://github.com/Automattic/jetpack/pull/22062/files#r766771565

anomiex committed:
avoid warnings when widgets is not an array

See https://github.com/Automattic/jetpack/pull/22062/files#r766773643

jeherve commented on projects/plugins/jetpack/modules/widgets/migrate-to-core/image-widget.php:
```diff
@@ -176,9 +185,11 @@ function jetpack_migrate_image_widget() {
 		}
 
 		foreach ( $sidebars_widgets as $sidebar => $widgets ) {
+			$key = array_search( "image-{$id}", $widgets, true );
```
👍 db601e7528d102ac5aef407efc877a00d2c9d2f0

jeherve committed:
Cast number as an int so we can do strict comparisons.

See https://github.com/Automattic/jetpack/pull/22062/files#r766777630

jeherve commented on projects/plugins/jetpack/modules/widgets/my-community.php:
```diff
@@ -68,14 +68,14 @@ function enqueue_style() {
 
 		$number = isset( $instance['number'] ) ? $instance['number'] : 10;
-		if ( ! in_array( $number, array( 10, 50 ) ) ) {
+		if ( ! in_array( $number, array( 10, 50 ) ) ) { // phpcs:ignore WordPress.PHP.StrictInArray.MissingTrueStrict
```
👍 0d453bd02ab96c1713664f3315745bbf430ae7d2

anomiex committed:
Prefer strict comparison, we know that's going to be an int

See https://github.com/Automattic/jetpack/pull/22062/files#r766775597

jeherve commented on projects/plugins/jetpack/modules/widgets/my-community.php:
```diff
@@ -131,15 +131,15 @@ function form( $instance ) {
 
 		$instance['number'] = (int) $new_instance['number'];
-		if ( ! in_array( $instance['number'], array( 10, 50 ) ) ) {
+		if ( ! in_array( $instance['number'], array( 10, 50 ) ) ) { // phpcs:ignore WordPress.PHP.StrictInArray.MissingTrueStrict
```
👍 977b3ccaa8cb5e555d388f6d3b77efb81644f08f

jeherve committed:
Do not escape widget titles

- Those come from WordPress, where they are not escaped for default widgets.
- Folks can use the `widget_title` filter to add HTML to titles, so let's not break that with extra escaping.

See https://github.com/Automattic/jetpack/pull/22062/files#r766779432

jeherve committed:
Add changelog for Search package

jeherve commented on projects/plugins/jetpack/modules/widgets/my-community.php:
```diff
@@ -180,10 +181,10 @@ function widget( $args, $instance ) {
 
 		if ( ! empty( $title ) ) {
-			echo $args['before_title'] . $title . $args['after_title'];
+			echo $args['before_title'] . esc_html( $title ) . $args['after_title']; // phpcs:ignore WordPress.Security.EscapeOutput.OutputNotEscaped
```
I'm not too sure what to make of this. We do escape the title in some of our widgets already, but not all. In core, they don't escape them indeed, but they do run them through `sanitize_text_field` on save. 

Maybe we shouldn't escape them, and assume folks using the `widget_title` filter may use the filter to add HTML to all widget titles.

I'll remove it from this one, and all others while I'm at it.
10edd7b10145c90eb33bcf0d6faccaa3301376f0

jeherve committed:
Update projects/plugins/jetpack/modules/widgets/rsslinks-widget.php

Co-authored-by: Brad Jorsch <anomiex@users.noreply.github.com>

jeherve committed:
Update projects/plugins/jetpack/modules/widgets/top-posts.php

Co-authored-by: Brad Jorsch <anomiex@users.noreply.github.com>

jeherve committed:
Update projects/plugins/jetpack/modules/widgets/top-posts.php

Co-authored-by: Brad Jorsch <anomiex@users.noreply.github.com>

jeherve committed:
Update projects/plugins/jetpack/modules/widgets/upcoming-events.php

Co-authored-by: Brad Jorsch <anomiex@users.noreply.github.com>

anomiex committed:
Switch to a for loop

See https://github.com/Automattic/jetpack/pull/22062/files#r766851917

jeherve commented:
Thanks for the review. I think I've addressed all your comments, it should be ready for another go!

anomiex commented on projects/plugins/jetpack/modules/widgets/migrate-to-core/gallery-widget.php:
```diff
@@ -92,81 +107,93 @@ function jetpack_migrate_gallery_widget_is_importable( $widget ) {
+		$key = is_array( $widgets ) ? array_search( "gallery-{$id}", $widgets, true ) : false;
+
 		if (
 			is_array( $widgets )
-			&& false !== ( $key = array_search( "gallery-{$id}", $widgets, true ) )
+			&& false !== $key
 		) {
```
No need to check `is_array()` again now.
```suggestion
		if ( false !== $key ) {
```

jeherve replied:
👍 b7c781e90394caa4c851f6a770e99d7648721e90

anomiex commented on projects/plugins/jetpack/modules/widgets/migrate-to-core/image-widget.php:
```diff
@@ -176,9 +185,11 @@ function jetpack_migrate_image_widget() {
+			$key = is_array( $widgets ) ? array_search( "image-{$id}", $widgets, true ) : false;
+
 			if (
 				is_array( $widgets )
-				&& false !== ( $key = array_search( "image-{$id}", $widgets, true ) )
+				&& false !== $key
 			) {
```
No need to check `is_array()` again now.
```suggestion
		if ( false !== $key ) {
```

jeherve replied:
👍 b7c781e90394caa4c851f6a770e99d7648721e90

anomiex reviewed (commented):
Two minor suggestions, otherwise looks good

jeherve committed:
Merge remote-tracking branch 'origin/master' into update/phpcs-widgets-2

anomiex committed:
No need to check for presence of an array anymore

See:
- https://github.com/Automattic/jetpack/pull/22062/files#r766924093
- https://github.com/Automattic/jetpack/pull/22062/files#r766924769

jeherve committed:
Merge remote-tracking branch 'origin/master' into update/phpcs-widgets-2

jeherve committed:
Update Search package version

jeherve commented:
I'm afraid this is going to need one last review, I had to bump a package's version.

jeherve commented:
r236617-wpcom
